File: bank_loan/erlang/money.erl
%%%-------------------------------------------------------------------
%%% @author odira
%%% @copyright (C) 2019, <COMPANY>
%%% @doc
%%%
%%% @end
%%% Created : 12. Jun 2019 11:18 PM
%%%-------------------------------------------------------------------
-module(money).
-author("odira").

%% API
-export([start/0]).

start() ->
  {_, Customers} = file:consult("customers.txt"),
  {_, Banks} = file:consult("banks.txt"),
  ets:new(toprint_customer, [set,named_table]),
  ets:new(toprint_customerNE, [set, named_table]),
  ets:new(toprint_bank, [set, named_table]),
  io:format("~n ** Customers and loan objectives ** ~n~n"),
  lists:foreach(fun list_customers/1, Customers),
  io:format("~n ** Banks and Financial resources ** ~n~n"),
  lists:foreach(fun list_banks/1, Banks),
  master(Banks, Customers).

list_customers({Customer, Total_funds}) ->
  io:format("~p: ~p~n",[Customer, Total_funds]).

list_banks({Bank, Total_loan}) ->
  io:format("~p: ~p~n",[Bank, Total_loan]).

master(Banks, Customers) ->
  io:format("~n"),
  create_bank_process(Banks),
  create_customer_process(Customers),

  self() ! handle_request([Banks], [Customers]).


create_bank_process([]) ->
  io:format("");
%%  io:format("~n~n Processes: ~n ~p ~n~n", [registered()]);

create_bank_process([Current_Bank|Other_Banks]) ->
  Current_Bank_Name = element(1, Current_Bank),
  register(Current_Bank_Name, spawn(bank, start, [self(), Current_Bank_Name, element(2, Current_Bank)])),
  create_bank_process(Other_Banks).

create_customer_process([]) ->
  io:format("");
%%  io:format("~n~n Processes: ~n ~p ~n~n", [registered()]);

create_customer_process([Current_Customer|Other_Customers])  ->
  Current_Customer_Name = element(1, Current_Customer),
  register(Current_Customer_Name, spawn(customer, start, [self(), Current_Customer, element(2, Current_Customer)])),
  create_customer_process(Other_Customers).

print_list([]) ->
  io:format("");

print_list([Head|Tail]) ->
  io:format("~p has reached the objective of ~p dollar(s). Woo Hoo! ~n", [element(1, Head), element(2, Head)]),
  print_list(Tail).

print_list_bank([]) ->
  io:format("");

print_list_bank([Head|Tail]) ->
  io:format("~p has ~p dollar(s) remaining. ~n", [element(1, Head), element(2, Head)]),
  print_list_bank(Tail).

print_list_ne([]) ->
  io:format("");

print_list_ne([Head|Tail]) ->
  io:format("~p was only able to borrow ~p dollar(s). Boo Hoo! ~n", [element(1, Head), element(2, Head)]),
  print_list_ne(Tail).

handle_request([Banks], [Customers])  ->
  receive
    {Pid, "bank_request"} ->
      Pid ! {self(), [Banks]},
%%      io:format("bank request handled for ~p ~n", [Pid]),
      handle_request([Banks], [Customers]);
    {Pid, "loan_request", Customer, Loan_amount, Bank} ->
      io:format("~p requests a loan of ~p dollar(s) from ~p ~n", [Customer, Loan_amount, Bank]),
      handle_request([Banks], [Customers]);
    {Pid, "loan_approved", Customer, Loan_amount, Bank} ->
      io:format("~p approves a loan of ~p dollars from ~p ~n", [Bank, Loan_amount, Customer]),
      handle_request([Banks], [Customers]);
      {Pid, "loan_denied", Customer, Loan_amount, Bank} ->
      io:format("~p denies a loan of ~p dollars from ~p ~n", [Bank, Loan_amount, Customer]),
      handle_request([Banks], [Customers]);
      {Pid, "goal_reached", Customer, Loan} ->
%%      io:format("~p has reached the objective of ~p dollar(s). Woo Hoo! ~n", [Customer, Loan]),
      handle_request([Banks], [Customers]);
      {Pid, "Notifications", "customer", [Message]} ->
      add_message(Message),
      handle_request([Banks], [Customers]);
    {Pid, "Notifications", "bank", Bank, Fund} ->
      add_message(Bank, Fund),
      handle_request([Banks], [Customers]);
      {Pid, "Notifications", "customer_notenough", Name, Fund} ->
        add_message_ne(Name, Fund),
        handle_request([Banks], [Customers])

      after 3000 ->
          print_list(ets:tab2list(toprint_customer)),
          print_list_ne(ets:tab2list(toprint_customerNE)),
          print_list_bank(ets:tab2list(toprint_bank))
  end.

add_message(Message) ->
%%  io:format("~p", [Message]),
  ets:insert(toprint_customer, [Message]).

add_message(Bank, Fund) ->
  Message = {Bank, Fund},
  ets:insert(toprint_bank, Message).

add_message_ne(Name, Fund) ->
  Customer = {Name, Fund},
  ets:insert(toprint_customerNE, Customer).

File: bank_loan/money.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.ThreadLocalRandom;

class Customer implements Runnable{
    String name;
    volatile Integer loan;
    HashMap<String, Bank> banks;
    Thread thread;
    Random rand = new Random();
    ArrayList<String> keysAsArray;


    volatile Integer recievedLoan;
    Integer loanCounter;
    Integer randomLoan;
    boolean approvalStatus = false;

    Customer(String Name, Integer Loan, HashMap Banks){
        name = Name;
        loan = Loan;
        loanCounter = Loan;
        banks = Banks;
        keysAsArray = new ArrayList<>(Banks.keySet());


        thread = new Thread(this, name);
        thread.start();

    }

    public void run(){
        recievedLoan = 0;
        boolean check = false;

            while (loan - recievedLoan != 0 && !keysAsArray.isEmpty()) {
                if (check == true){
                    if (loan - recievedLoan == 1){
                        randomLoan = 1;
                    }else if (loan - recievedLoan > 1){
                        randomLoan = ThreadLocalRandom.current().nextInt(1, loan - recievedLoan);
                    }
                    else{
                        break;
                    }
                }else{
                    randomLoan = ThreadLocalRandom.current().nextInt(1, 50);
                }




                boolean checkRandom = (randomLoan + recievedLoan) > loan;
                if (checkRandom) {
                    check = true;
                    continue;
                } else {
                    String key = keysAsArray.get(rand.nextInt(keysAsArray.size()));
                    if (this.banks.get(key).loanRequest(this, randomLoan)) {
                        updateRecievedLoan(randomLoan);
                        loanCounter = loanCounter - recievedLoan;
                        check = false;
                    } else {
                        keysAsArray.remove(key);
                    }
                }
            }


            if (loan - recievedLoan == 0) {
                money.addFinalMessage(Thread.currentThread().getName() + " has reached the objective of " + recievedLoan + " dollar(s). Woo Hoo!");
            } else {
                money.addFinalMessage(Thread.currentThread().getName() + " was only able to borrow " + recievedLoan + " dollar(s). Boo Hoo!");
            }
    }

    private void updateRecievedLoan(Integer randomLoan) {
        this.recievedLoan += randomLoan;
    }


}

















class Bank implements Runnable{
    String name;
    Integer funds;
    Thread thread;
    static boolean flag = true;
    String message = null;
    Random rand = new Random();

    Bank(String Name, Integer Funds){
        name = Name;
        funds = Funds;
        thread = new Thread(this, name);
        thread.start();

    }

    public void run(){
        try {
            Thread.sleep(5000);


//            while (flag){
//                if (message != null){
//                    money.printMessage(message);
//                    message = null;
//                }
//            }

        } catch (InterruptedException e){
            e.printStackTrace();
        }

        money.addFinalMessage(Thread.currentThread().getName() + " has " + funds + " dollar(s) remaining.");
        //System.out.println(name + "exiting");
    }

    public synchronized boolean loanRequest(Customer customer, Integer Amount){
        money.printMessage(Thread.currentThread().getName() + " requests a loan of "+ Amount + " dollar(s) from " + this.name);
        boolean result = false;
        try {
            Thread.sleep(rand.nextInt((100 - 10) + 1) + 10);
            if (funds - Amount >= 0){
                money.printMessage(this.name + " approves a loan of " + Amount + " dollars from " + Thread.currentThread().getName());
                updateFunds(funds - Amount);
                result = true;
            }else {
                money.printMessage(this.name + " denies a loan of " + Amount + " dollars from " + Thread.currentThread().getName());
                result = false;
            }

        } catch (Exception e) {
            e.printStackTrace();
        }



        return result;
    }

    public void updateFunds(Integer newFunds){
        this.funds = newFunds;
    }

}































public class money extends Thread{
    static HashMap<String, Integer> banks = new HashMap<>();
    static HashMap<String, Integer> customers = new HashMap<>();
    static HashMap<String, Bank> bankObject = new HashMap<>();
    static ArrayList<String> finalMessages = new ArrayList<>();
    static ArrayList<Thread> customerThreads = new ArrayList<>();
    static ArrayList<Thread> bankThreads = new ArrayList<>();
    static Thread mainThread;



    public static synchronized void printMessage(String Message){
        System.out.println(Message);
    }

    public static synchronized void addFinalMessage(String Message){
        finalMessages.add(Message);
    }

    public synchronized void printFinalMessages(){
        for (String finalMessage : finalMessages){
            System.out.println(finalMessage);
        }
    }

    public static void readFiles() throws Exception {
        FileReader customerFile = new FileReader("src/customers.txt");
        FileReader bankFile = new FileReader("src/banks.txt");
        BufferedReader reader = new BufferedReader(customerFile);

        System.out.println("\n ** Customers and loan objectives ** \n");
        String line = reader.readLine();
        while(line != null)
        {
            line = line.replace("{", "");
            line = line.replace("}", "");
            line = line.replace(".", "");
            String[] tempLine = line.split(",");
            customers.put(tempLine[0],Integer.parseInt(tempLine[1]));
            System.out.println(line.replace(",",": "));
            line = reader.readLine();

        }

        reader = new BufferedReader(bankFile);
        System.out.println("\n ** Banks and financial resources ** \n");
        line = reader.readLine();
        while(line != null)
        {
            line = line.replace("{", "");
            line = line.replace("}", "");
            line = line.replace(".", "");
            String[] tempLine = line.split(",");
            banks.put(tempLine[0],Integer.parseInt(tempLine[1]));
            System.out.println(line.replace(",",": "));
            line = reader.readLine();

        }
        System.out.println("\n\n");
        reader.close();


    }


    public static void main (String args[]) throws InterruptedException {

        try{
            readFiles();
            new money().start();
        } catch (Exception e){
            e.printStackTrace();
        }

    }

    @Override
    public void run() {
        for (Map.Entry<String, Integer> entry : banks.entrySet()){
            Bank bank = new Bank(entry.getKey(), entry.getValue());
            bankObject.put(entry.getKey(), bank);
            bankThreads.add(bank.thread);

        }

        for (Map.Entry<String, Integer> entry : customers.entrySet()){
            Customer customer = new Customer(entry.getKey(), entry.getValue(), bankObject);
            customerThreads.add(customer.thread);
        }



        for (Thread t : customerThreads){
            try {
                t.join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        for (Thread t : bankThreads)
        {
            try {
                t.join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        printFinalMessages();

    }
}
